### DISCUSSION TITLE
Unable to boot from local disk unless it's part of the boot order

### DISCUSSION THREAD

**Discussion by thannaske** [ipxe/ipxe#1034] [General]
I'm trying to boot a server from local disk via iPXE. The reason I'm using iPXE is for dynamic control of what the server shall boot. I don't want the local disk in the normal boot order so I have full control over the actual boot process. If my hypervisor is disconnected from the network, I'd rather have the machine stuck in a boot loop than booting from local disk without fetching my iPXE script.

The configuration looks like this:

```
  <os>
    <type arch='x86_64' machine='pc-i440fx-jammy'>hvm</type>
    <boot dev='network'/>
    <bios useserial='yes' rebootTimeout='60'/>
    <smbios mode='sysinfo'/>
  </os>
```

However, when I try booting the server like this, I get the following error:

```
menu: 49 bytes [script]
Booting from SAN device 0×80
Boot from SAN device 0x80 failed: Input/output error (https://ipxe.org/err/1d852039)

Could not boot image: Input/output error (https://ipxe.org/err/1d852039)
No more network devices
```

When I now edit the libvirt domain definition like this, the boot process works:

```
  <os>
    <type arch='x86_64' machine='pc-i440fx-jammy'>hvm</type>
    <boot dev='network'/>
+ <boot dev='hd'/>
    <bios useserial='yes' rebootTimeout='60'/>
    <smbios mode='sysinfo'/>
  </os>
```

The server is then actually booting from iPXE, it never touches the second boot device. It only works because that boot device is listed alongside the network boot option. However, I can't leave it like this, because when the hypervisor looses connectivity or my iPXE backend isn't responding the server would just boot from its local disk, although it might not be what I want in that situation.

Did anybody experience something like this before?
The server is running libvirtd (libvirt) 8.0.0 and iPXE is installed as iPXE 1.21.1+git-20220113.fbbdc3926-0ubuntu1 (Ubuntu 22.04 LTS default packages).

I found out that the issue **doesn't** exist under Ubuntu 20.04 where libvirtd ships in version 6.0.0 and iPXE in version 1.0.0+git-20190109.13354c4-0ubuntu3.2

**Comment by NiKiZe** [comment]
Officially iPXE does "not support booting from local disk".
There is several discussions on this topic, try the search function.

The recommended/supported way of doing disk boot in iPXE is to do a exit which should give bios the option to boot the next boot device.

In pcbios you can do sanboot 0x80 which is the location of the first disk. If that fails it should mean that there is no disk at that location (primary ide master in older computers, which the 440 is) or that the disk don't have valid boot code.

You do however get s url in the message (iPXE always does) and it has a good advice, use latest version of iPXE, 2022 is quite old. One good option is to always chain to a newer version of iPXE, regardless of what Qemu ships with (unless you replace the ROM used by Qemu directly)

**Comment by mcb30** [comment]
Follow the link that iPXE gives you in the error message: https://ipxe.org/1a852039

As stated on that page:
| This error no longer exists in the iPXE source code. You should try using the [latest version](https://ipxe.org/download) of iPXE.

Try with the latest version of iPXE and see if the problem still exists.  If it does, then report the error that you get when you try to use that latest version.

**Reply by thannaske** [reply]
Thank you very much for your answer. I already built a new ROM for the E1000 network adapter I use in my guest definition. I tried to place the file under `/usr/lib/ipxe/qemu` where I found a bunch of those ROMs. However, it doesn't seem to load my custom-built ROM but rather the default one that ships with Ubuntu 22.04 LTS.

Can you give me a hint where I need to put those ROMs in order to get them loaded or whether I need to change some configuration with qemu/libvirt? I couldn't find anything regarding this topic in the documentation in either documentation so please forgive me my dumb question 😅

Edit:
I just found out that the error message was pasted wrong. The actual error URL reads: https://ipxe.org/err/1d852039

**Reply by NiKiZe** [reply]
So BIOS can't read MBR https://github.com/ipxe/ipxe/blob/master/src/arch/x86/interface/pcbios/int13.c#L1392

How does your disk section look like in the libvirt config? Please include all disks and only censor parts of filenames and not full paths.

**Reply by thannaske** [reply]
There's only a single disk and a single network interface in my domain configuration:

```xml
  [...]
  <devices>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2' cache='none'/>
      <source file='/STORAGE/server-4-boot.qcow2'/>
      <target dev='sda' bus='scsi'/>
      <address type='drive' controller='0' bus='0' target='0' unit='0'/>
    </disk>
    <controller type='scsi' index='0' model='virtio-scsi'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x05' function='0x0'/>
    </controller>
    [...]
    <interface type='network'>
      <mac address='52:54:00:4c:d1:dc'/>
      <source network='default'/>
      <model type='e1000'/>
      <acpi index='1'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </interface>
    [...]
  </devices>
```

At the top, I defined:

```xml
<os>
  <boot dev='network'/>
</os>
```

When I alter this section like this, it works immediately:
```xml
<os>
  <boot dev='network'/>
  <boot dev='hd'/>
</os>
```

But I don't want to let the server boot from disk by default, it's only the exception that the server shall boot directly from its disk and if it does, I need full control over it with a dynamic script that instructs the server to do so.

**Reply by NiKiZe** [reply]
So sanboot 0x80 works when you have it in the boot order list?
If so I assume that Qemu blocks access to the MBR thru INT13 when it's not in list.

Not much iPXE can do. Try different version of Qemu and see if something is changed, boot in EFI instead or maybe work around the issue thru a chained bootloader, such as syslinux or grub

**Reply by thannaske** [reply]
Yes, correct. If it's in the boot order, `sanboot 0x80` works as intended. As it works in the older QEMU versions, I indeed assume that something changed in the versions. I'll try to work around it.

Thanks a lot for your valuable input!
